egressgw: retry getIdentityLabels on failure

jibi opened the pull request:
when the manager fails to resolve the identity of an endpoint to a set of labels, instead of discarding the endpoint update, enqueue the event and retry identity labels resolution following the same exponential backoff logic used by the identity allocator

### testing 
I tested this locally by forcing the identity resolution to randomly fail:

```diff
diff --git a/pkg/egressgateway/manager.go b/pkg/egressgateway/manager.go
index ed7f19399c..0b867b2d12 100644
--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -7,6 +7,7 @@ import (
        "container/heap"
        "context"
        "fmt"
+       "math/rand"
        "net"
        "sort"
        "time"
@@ -171,6 +172,10 @@ func NewEgressGatewayManager(p Params) *Manager {
 // getIdentityLabels waits for the global identities to be populated to the cache,
 // then looks up identity by ID from the cached identity allocator and return its labels.
 func (manager *Manager) getIdentityLabels(securityIdentity uint32) (labels.Labels, error) {
+       if rand.Intn(16) > 0 {
+               return nil, fmt.Errorf("fake error")
+       }
+
```

```sh
➜  cilium git:(pr/jibi/egressgw-retry-identity) ks logs cilium-2dc5v --timestamps | grep subsys=egress | grep coredns-5d78c9869d-44vlf
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
2023-06-23T10:22:36.698593375Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:37.590584472Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:42.630015783Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:44.710650728Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:46.871488814Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:52.190670670Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:55.113841789Z level=warning msg="Failed to get identity labels for endpoint, will retry." error="fake error" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
2023-06-23T10:22:58.952487945Z level=debug msg="Added CiliumEndpoint" k8sEndpointName=coredns-5d78c9869d-44vlf k8sNamespace=kube-system subsys=egressgateway
```

Fixes: #26158 (hopefully :crossed_fingers:)

bimmlerd reviewed (approved):
CODEOWNERS change lgtm

giorio94 commented on pkg/backoff/backoff.go:
```diff
@@ -201,3 +206,8 @@ func (b *Exponential) Duration(attempt int) time.Duration {
 	return t
 }
+
+// CurrentDuration returns the wait duration for the current number of attempt
```
```suggestion
// CurrentDuration returns the wait duration for the current number of attempts
```

Nit.

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -149,6 +156,7 @@ func NewEgressGatewayManager(p Params) *Manager {
+			manager.retryEndpointIdentityLabelsResolution(ctx)
 			return nil
 		},
 		OnStop: func(hc hive.HookContext) error {
```
Nit. According to the [hive guidelines](https://docs.cilium.io/en/latest/contributing/development/hive/#guidelines), the stop hook should block until goroutines have terminated.

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -190,6 +198,35 @@ func (manager *Manager) runReconciliationAfterK8sSync(ctx context.Context) {
+			case <-ticker.C:
+				manager.Lock()
+				queueLen := manager.endpointEventRetryQueue.Len()
+				manager.Unlock()
```
There's the possibility of a race condition, given that the lock is released here, but then the heap is accessed again below to pop one item.

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -190,6 +198,35 @@ func (manager *Manager) runReconciliationAfterK8sSync(ctx context.Context) {
+				}
+
+				endpointEventRetry := heap.Pop(&manager.endpointEventRetryQueue).(*endpointEventRetry)
+				endpointEventRetry.backoff.Wait(ctx)
```
There's the possibility that this wait blocks also other "higher priority" retries. Let's say for instance that one endpoint has already been retried a few times, and needs to block here for a long period (t1). A new endpoint which is queued immediately afterwards (with a smaller wait period) will have to wait for t1 + t2, rather than just t2. The same happens in case multiple endpoints are queued, the latter will have to wait for n*t, although its own waiting period should have been only t.

More generally, I'd suggest to use a working queue rather than a custom implementation, given that it already automatically handles all these aspects.

jibi replied:
ack, let me take a look at working queues then :+1:

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -190,6 +198,35 @@ func (manager *Manager) runReconciliationAfterK8sSync(ctx context.Context) {
+
+				endpointEventRetry := heap.Pop(&manager.endpointEventRetryQueue).(*endpointEventRetry)
+				endpointEventRetry.backoff.Wait(ctx)
+				manager.addEndpoint(endpointEventRetry.endpoint, endpointEventRetry.backoff)
```
There's the possibility that an endpoint which is added here after the wait period has already been deleted in parallel.

jibi committed:
egressgw: retry getIdentityLabels on failure

when the manager fails to resolve the identity of an endpoint to a set
of labels, instead of discarding the endpoint update, enqueue the event
and retry identity labels resolution following the same exponential
backoff logic used by the identity allocator

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: move endpoint.Networking.Addressing check

inside getEndpointMetadata as that's where the rest of the checks for
the received CiliumEndpoint event live

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -265,23 +355,63 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	// remove the ID from the workqueue in case an older revision of the
+	// same CiliumEndpoint was already present
+	manager.endpointEventRetryQueue.Forget(id)
+	manager.endpointEventRetryQueue.Done(id)
```
It is a bit strange seeing these operations being performed outside of the reconcile loop, but I cannot spot anything that could go wrong with them. Still, I wonder if it could make sense to just add the entry to the pending events and push the ID in the queue, leaving the reconcile loop to trigger `addEndpoint`.

jibi replied:
I think this should stay here as it has nothing to do with the reconciliation logic. More in general, the way the manager is structured is the following:
* we have a bunch of event handlers called by the k8s watchers (for policies, endpoints and nodes events) which simply parse the events and update the copies of these resources stored in the manager
* and then we have the reconciliation logic which goes through these objects and ensure the state of the node is in sync with these resources

because of this I think parsing an endpoint event (and retrying because we failed to resolve identity -> labels) should not be part of the reconciliation logic

giorio94 replied:
Yeah, I agree, and I think we might be referring to different things when mentioning "reconciliation logic". In particular, I was meaning the `retryEndpointIdentityLabelsResolution` loop, and not `manager.reconcile()` (I agree that the latter is independent and should not handle the endpoint parsing for symmetry with the other handlers). 

Still, my comment was mostly a nit, and I'm also fine with the current version.

jibi replied:
Ah, I see what you mean now :+1: 
I changed the `On{Update,Delete}Endpoint` handlers to only add and remove items from the workqueue/pending events map, and let the retry loop do the rest (so call `addEnpoint` and deal with re-adding back items or marking them as done)

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -265,23 +355,63 @@ func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
 
+	delete(manager.pendingEndpointEvents, id)
+	manager.endpointEventRetryQueue.Forget(id)
+	manager.endpointEventRetryQueue.Done(id)
```
Same here about just letting the event to be consumed by the reconcile loop rather than removing it explicitly.

giorio94 reviewed (approved):
The workqueue implementation looks good to me.
I wonder whether it could make sense to leverage it also during insertion for increased uniformity. WDYT?

giorio94 commented on pkg/egressgateway/manager.go:
```diff
@@ -190,6 +218,47 @@ func (manager *Manager) runReconciliationAfterK8sSync(ctx context.Context) {
+// retryEndpointIdentityLabelsResolution spawns a goroutine that retries to
+// add/update an endpoint for which we failed to resolve the identity to a set
+// of labels
+func (manager *Manager) retryEndpointIdentityLabelsResolution(ctx context.Context, wg *sync.WaitGroup) {
```
Nit. The name of this function (and the comment) could be generalized now that handles all upsert events.

giorio94 reviewed (approved):
Looks great to me!
